### PR TITLE
[doc] Hyperparameter Tuning Gallery Entry

### PR DESCRIPTION
**Pull request by zhu-eric**
Add comments to plot_hyperparameter.py

### CODE REVIEW

File: doc/examples/plot_hyperparameter.py [hunk 1]
```diff
@@ -99,10 +111,6 @@ def train(model, optimizer, train_loader, device=torch.device("cpu")):
 
 
 def test(model, test_loader, device=torch.device("cpu")):
-    """Checks the validation accuracy of the model.
```
Comment (richardliaw) [inline doc/examples/plot_hyperparameter.py]: Let's keep this

File: doc/examples/plot_hyperparameter.py [hunk 2]
```diff
@@ -14,6 +14,11 @@
     tuning library built using Ray's Actor API.
 
 .. _`Tune`: https://ray.readthedocs.io/en/latest/tune.html
+
+Setup: Dependencies
+-------------------
+First, import some dependencies and define functions to generate
+random hyperpaameters and retrieve data.
```
Comment (richardliaw) [inline doc/examples/plot_hyperparameter.py]:
```suggestion
random hyperparameters and retrieve data.
```

File: doc/source/conf.py [hunk 3]
```diff
@@ -377,6 +377,7 @@ def update_context(app, pagename, templatename, context, doctree):
     """Update the page rendering context to include ``feedback_form_url``."""
     context['feedback_form_url'] = feedback_form_url(app.config.project,
                                                      pagename)
+# see also http://searchvoidstar.tumblr.com/post/125486358368/making-pdfs-from-markdown-on-readthedocsorg-using
```
Comment (richardliaw) [inline doc/source/conf.py]: duplicate?

File: doc/examples/plot_hyperparameter.py [hunk 4]
```diff
@@ -141,13 +169,22 @@ def evaluate_hyperparameters(config):
...
 # Randomly generate sets of hyperparameters and launch a task to test each set.
 for i in range(num_evaluations):
     hyperparameters = generate_hyperparameters()
     accuracy_id = evaluate_hyperparameters.remote(hyperparameters)
     remaining_ids.append(accuracy_id)
     hyperparameters_mapping[accuracy_id] = hyperparameters
 
+###########################################################################
+# Process each hyperparameter and corresponding accuracy in the order that 
+# they finish training and testing the neural network to store the
```
Comment (richardliaw) [inline doc/examples/plot_hyperparameter.py]:
```suggestion
# they finish to store the 
```

File: doc/examples/plot_hyperparameter.py [hunk 5]
```diff
@@ -141,13 +169,22 @@ def evaluate_hyperparameters(config):
...
 for i in range(num_evaluations):
     hyperparameters = generate_hyperparameters()
     accuracy_id = evaluate_hyperparameters.remote(hyperparameters)
     remaining_ids.append(accuracy_id)
     hyperparameters_mapping[accuracy_id] = hyperparameters
 
+###########################################################################
+# Process each hyperparameter and corresponding accuracy in the order that 
+# they finish training and testing the neural network to store the 
+# hyperparameters with the best accuracy
```
Comment (richardliaw) [inline doc/examples/plot_hyperparameter.py]:
```suggestion
# hyperparameters with the best accuracy.
```

File: doc/examples/plot_hyperparameter.py [hunk 6]
```diff
@@ -141,13 +169,22 @@ def evaluate_hyperparameters(config):
 # hyerparameters used for that experiment.
 hyperparameters_mapping = {}
 
+###########################################################################
+# Compute accuracy for different randomly generated hyperparameters, and
```
Comment (richardliaw) [inline doc/examples/plot_hyperparameter.py]: Launch asynchronous parallel tasks for evaluating different hyperparameters. `accuracy_id` is an ObjectID that acts as a handle to the remote task. It is used later to fetch the result of the task when the task finishes.

File: doc/examples/plot_hyperparameter.py [hunk 7]
```diff
@@ -130,6 +150,14 @@ def evaluate_hyperparameters(config):
     train(model, optimizer, train_loader)
     return test(model, test_loader)
 
+#######################################################################
+# Synchronous Evaluation of Randomly Generated Hyperparameters
+# ------------------------------------------------------------
+#
+# We will create multiple sets of random hyperparameters for our neural
+# network that will be trained in parallel. These trained networks will
```
Comment (richardliaw) [inline doc/examples/plot_hyperparameter.py]:
```suggestion
# network that will be evaluated in parallel. These trained networks will
```

Comment (richardliaw) [reply doc/examples/plot_hyperparameter.py]: Also, I think the last sentence "tested for accuracy..." should be moved to the previous block

**Review by richardliaw — COMMENTED** [review]
Left some more comments - we should be good to merge after this

File: doc/examples/plot_hyperparameter.py [hunk 8]
```diff
@@ -65,6 +70,13 @@ def get_data_loaders(batch_size):
         shuffle=True)
     return train_loader, test_loader
 
+#######################################################################
+# Setup: Defining the Neural Network
+# ----------------------------------
+#
+# We define a small neural network to use in training. In addition, 
+# we created methods to train and test this neural network
```
Comment (richardliaw) [inline doc/examples/plot_hyperparameter.py]:
```suggestion
# we created methods to train and test this neural network.
```

**Review by richardliaw — APPROVED** [review]
